### ISSUE TITLE
observatory of any and all chain forks

### ISSUE THREAD

**Issue by zookozcash** [zcash/zcash#1925]
This inspects the network to detect a chain fork in which different nodes are following different forks for more than a few [*] blocks. If it detects such a chain fork, it alerts the user.

Ideally, I think, it should rely on [the block observatory](https://github.com/zcash/zcash/issues/1922) and should be inspecting the "catalog of all blocks" produced by that.

[*] 6, because https://github.com/zcash/zcash/issues/952

**Comment by str4d** [comment]
Status update: I decided against using the DNS seeder as a starting point, because it didn't implement any block-parsing logic at all, and therefore could not collect the types of information necessary to detect chain forks. Instead, I have forked https://github.com/jgarzik/pynode which implements a basic Bitcoin node and collects blocks from a chosen peer.

I have [updated it](https://github.com/str4d/pynode/tree/bitcoinlib-0.7.0) to work with the latest `python-bitcoinlib`, which I am porting to Zcash. All that remains is to fix the bugs in my JoinSplit parser, and then we can use `pynode` as-is to collect blocks from a single peer. Thus the first trivial version of a general chain fork detector would be to run several `pynode` instances, and then manually compare their blocks.

Tomorrow, I will finish chasing down the parsing bugs, and work on hacking multi-peer connectivity into `pynode`.

**Comment by str4d** [comment]
Parsing bugs have been resolved, and the only remaining issue is that transaction signature validation isn't working. But that isn't necessary for a chain fork detector, so I've worked around it for now.

**Comment by nathan-at-least** [comment]
Bumped from 1.0.9, mainly because I don't know how to evaluate if I can close this ticket, and because the work is decoupled from our release, and we're over-capacity for 1.0.9.

**Comment by nathan-at-least** [comment]
I propose closing this ticket with a worse-is-better solution (thus unblocking it from the Block Observatory):

- we set up multiple detector nodes (Bonus: running a variety of release versions, #2483).
- we have a monitoring process collect the `getchaintips` output from all of these nodes.
- that monitoring process issues alerts whenever a `branchlen >= K` is observed. Let's say K=6 for now. (Because `getchaintips` is persistent, we need to be able to explicitly silence alerts for already observed branches.)
- that monitoring process should feed all results into a metrics system.

I believe we already have all 4 of these implemented. If so we could close the ticket. I'd like a review of all four pieces, first, though. In particular, I believe the alerting system and metrics aggregation are buggy and need improvement (which should be specified in separate tickets).

**Comment by nathan-at-least** [comment]
Ah, I just realized when @zookozcash says "the user" in the description I thought "ZcashCo operations team" but maybe he meant "any user of Zcash". That totally changes things. I propose we bite off the first goal (internal alerting system) as the first milestone, then followup with a public-facing interface (and cloneable open code others can deploy).

So I just renamed this, and will file a new ticket for the public facing service…

**Comment by nathan-at-least** [comment]
@str4d I don't quite follow your approach. I imagined polling the result of `getchaintips` from multiple nodes. How does your approach compare to that? (It sounds more involved to me.)
